hazelcast/src/test/java/com/hazelcast/jet/impl/processor/AsyncTransformUsingServicePTest.java

```diff
@@ -172,6 +176,34 @@ public void test_forwardMultipleWatermarksWithoutItems() {
                 .expectOutput(asList(wm(10, (byte) 0), wm(5, (byte) 1), wm(0, (byte) 2)));
     }
 
+    @Test
+    public void test_completedFutures_sameElement() {
+        TestSupport
+                .verifyProcessor(getSupplier((ctx, item) ->  completedFuture(singleton(item + "-1"))))
+                .hazelcastInstance(instance())
+                .input(asList("a", "a", "a"))
+                .disableProgressAssertion()
+                .expectOutput(asList("a-1", "a-1", "a-1"));
+    }
+
+    @Test
+    public void test_completedFutures_sameElementInterleavedWithWatermark() {
+        TestSupport
+                .verifyProcessor(getSupplier((ctx, item) ->  completedFuture(singleton(item + "-1"))))
+                .hazelcastInstance(instance())
+                .input(asList("a", "a", wm(10), "a"))
+                .outputChecker((expected, actual) ->
+                        actual.equals(asList("a-1", "a-1", wm(10), "a-1"))
+                        // last item may be reordered with watermark in current implementation
+                        // but this is not required.
+                        || (!ordered && actual.equals(asList("a-1", "a-1", "a-1", wm(10))))
+                        // after snapshot restore watermark may be duplicated
+                        || (!ordered && actual.equals(asList("a-1", "a-1", wm(10), "a-1", wm(10))))
```

Comment:
@viliam-durina  I think that watermark duplication is allowed. Is that true?

Reply:
It's not allowed, they must be strictly monotonic.

Reply:
Watermark order with regards to items is preserved.
After a lot of debugging I confirmed that WM duplication is partially caused by the fact that we do not store `lastEmittedWms` in the snapshot. So if at the time of getting snapshot we have:

1. WM that was already emitted and no new WM was received
2. inflight item

then, after restore we get:

1. that old, already emitted, WM in `lastReceivedWms`
2. inflight item that assumes that it was received before any snapshot (`Long.MIN_VALUE`) even though to be strict it should be related to the WM (`lastReceivedWms` field is updated after restored items are sent for processing)
3. when inflight item completes, it triggers emission of WM from `lastReceivedWms` (check in `watermarkCount.isEmpty() && lastReceivedWms[i] > lastEmittedWms[i]`) which seems to make sense at least for other cases

I think this actually may be by design that given WM can be repeated after snapshot because, as the comment says:
> we restart at the oldest WM any instance was at the time of snapshot

Reply:
also the duplication does not occur during single execution, duplicated WM is emitted after restore from snapshot. But because of "we restart at the oldest WM any instance was at the time of snapshot" it probably can go backwards.

Still, emitting WM after unrelated item seems at least counter-intuitive.

Reply:
I debugged it too and yes, the observed output is legal. The duplicate WMs aren't emitted from a single instance of a processor (that wouldn't be legal), but from the processor after restore.

We save a simplified state to the snapshot. Before saving, we know exactly which item was received before which WM. But we can't save it exactly like this, because after restore, the WM can go back in at-least-once mode, and items can be re-partitioned, and we can't re-partition the `watermarkCounts`. I think we don't even have to save `lastReceivedWm` at all, we can just rely on the new WMs received after restore, but it's not too important to change, I can be wrong here.